tarofreighter/chain_porter.go

```diff
@@ -388,117 +390,304 @@ func (p *ChainPorter) waitForPkgConfirmation(pkg *OutboundParcelDelta) {
+	return nil
+}
+
+// transferReceiverProof retrieves the sender and receiver proofs from the
+// archive and then transfers the receiver's proof to the receiver. Upon
+// successful transfer, the asset parcel delivery is marked as complete.
+func (p *ChainPorter) transferReceiverProof(pkg *sendPackage) error {
+	ctx, cancel := p.CtxBlocking()
+	defer cancel()
+
+	// Retrieve sender proof from proof archive.
+	assetId := pkg.OutboundPkg.AssetSpendDeltas[0].WitnessData[0].PrevID.ID
+	senderProofBlob, err := p.cfg.AssetProofs.FetchProof(
+		ctx, proof.Locator{
+			AssetID:   &assetId,
+			ScriptKey: *pkg.SenderScriptKey.PubKey,
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("error fetching sender proof: %w", err)
+	}
+
+	// Retrieve receiver proof from proof archive.
+	locator := proof.Locator{
+		AssetID:   &assetId,
+		ScriptKey: pkg.ReceiverAddr.ScriptKey,
+	}
+	receiverProofBlob, err := p.cfg.AssetProofs.FetchProof(ctx, locator)
+	if err != nil {
+		return fmt.Errorf("error fetching receiver proof: %w", err)
+	}
+	receiverProof := &proof.AnnotatedProof{
+		Locator: locator,
+		Blob:    receiverProofBlob,
+	}
+
 	// If we have a proof courier instance active, then we'll launch a new
 	// goroutine to deliver the proof to the receiver.
 	//
 	// TODO(roasbeef): move earlier?
 	if p.cfg.ProofCourier != nil {
 		p.Wg.Add(1)
+
+		errorsChan := make(chan error, 1)
 		go func() {
 			defer p.Wg.Done()
 
 			// TODO(roasbeef): should actually also serialize the
 			// addr of the remote party here
-			addr := address.Taro{
-				Genesis:   receiverProofSuffix.Asset.Genesis,
-				ScriptKey: receiverProof.ScriptKey,
-				Amount:    receiverProofSuffix.Asset.Amount,
-			}
 			ctx, cancel := p.WithCtxQuitNoTimeout()
 			defer cancel()
-			err := p.cfg.ProofCourier.DeliverProof(
-				ctx, addr, receiverProof,
+
+			// Query parcel delivery log to ensure a sensible
+			// rate of parcel delivery attempts.
+			timestamps, err := p.cfg.ExportLog.QueryParcelDeliveryLog(
+				ctx, pkg.OutboundPkg.NewAnchorPoint,
 			)
 			if err != nil {
-				log.Errorf("unable to deliver proof: %v", err)
+				err = fmt.Errorf("unable to retrieve proof delivery "+
+					"logs: %w", err)
+				errorsChan <- err
+				return
+			}
+
+			// Determine whether the past receiver proof delivery
+			// attempts occurred far enough in the past to warrant
+			// a new set of delivery attempts. Otherwise, wait.
+			//
+			// Only wait if we have a non-zero number of past
+			// delivery attempts.
+			timeSinceLastAttempt := timeSinceLastDeliveryAttempt(timestamps)
+			backoffResetWait := p.cfg.ReceiverProofTransferConfig.BackoffResetWait
+			if len(timestamps) > 0 &&
+				timeSinceLastDeliveryAttempt(timestamps) <
+					backoffResetWait {
+
+				waitDuration := backoffResetWait - timeSinceLastAttempt
+				log.Infof("Waiting %v before attempting to "+
+					"deliver receiver proof to receiver",
+					waitDuration)
+				p.wait(waitDuration)
+			}
+			if err != nil {
```

Comment:
i think we are using the err from the last ` p.cfg.ExportLog.QueryParcelDeliveryLog` call

Reply:
You're right, thanks. The `err` was supposed to come from `p.wait`. Fixed in latest commits.